### ISSUE TITLE
Misleading output when BeforeSuite fails, and running in parallel, and Focussing a test

### ISSUE THREAD

**Issue by robdimsdale** [onsi/ginkgo#173]
Apologies for the convoluted title.

I have 10 integration tests which run in parallel and which share a `BeforeSuite`. The `BeforeSuite` validates the user-provided credentials, failing fast if they are invalid. When I `Focus` one of the tests and provide incorrect credentials, I see the expected failure in `BeforeSuite` but ginkgo prints the failure text 10 times. Ginkgo does also say that it only ran one test, which is expected.

In terms of the execution I think it's correctly running one test and skipping the other 9, but the output does not reflect this. See below:

```
Running in parallel across 7 nodes

STEP: Logging in


Failure [0.016 seconds]
[BeforeSuite] BeforeSuite
/Users/rhdz/go/src/github.com/robdimsdale/wundergo/integration_tests/wundergo_integration_suite_test.go:105

  Error - WL_ACCESS_TOKEN must be provided

[... repeat error message above 8 more times - omitted for brevity... ]

Failure [0.017 seconds]
[BeforeSuite] BeforeSuite
/Users/rhdz/go/src/github.com/robdimsdale/wundergo/integration_tests/wundergo_integration_suite_test.go:105

  Error - WL_ACCESS_TOKEN must be provided

  /Users/rhdz/go/src/github.com/robdimsdale/wundergo/integration_tests/wundergo_integration_suite_test.go:95
------------------------------

Ran 1 of 10 Specs in 0.027 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 9 Skipped
```

**Comment by robdimsdale** [comment]
Correction: i saw the message 7 times, not 10. I feel I've read research that suggests humans aren't very good at quickly counting the number of a set of objects once that number exceeds about four.

Anyway, it sounds like the `BeforeSuite` is running for each node, regardless of whether it actually has any tests to run. The fact there are ten tests is irrelevant.

This also manifests itself if there are fewer tests than nodes, though it's more likely to be encountered when skipping tests.

Perhaps there is a feature request here - "BeforeSuite should not run if there are no tests for that particular node to run" ?

I could look into a PR if you don't have time, though I'd probably need pointing in the right direction as 'm not familiar with the codebase.

**Comment by onsi** [comment]
A PR if you have time would be great!  Thanks @robdimsdale 

To make sure we're on the same page, first a mental model:

> When running in parallel across `n` nodes, the `ginkgo` cli launches `n` processes.  Each process is given an index, `n`, and the same random seed.  Based on this information each process divies up the tests to yield a specific subset of tests.  The process then runs `BeforeSuite`, the tests, and `AfterSuite`.

Now, the proposed behavior:

After filtering the tests, if a node discovers that there are **no tests** to run (perhaps they are all skipped? perhaps there are more nodes than tests?) the node should *not* run the `BeforeSuite` or `AfterSuite`.

There is one caveat.  If there is a `SynchronizedBeforeSuite` and/or `SynchronizedAfterSuite` then node 1 should *always* run the `Synchronized[Before|After]Suite` as it may be performing setup for the other nodes.

Most of these changes will be in [spec_runner.go](https://github.com/onsi/ginkgo/blob/master/internal/specrunner/spec_runner.go).  It will probably help to add a few integration tests as well.

One tricky piece will be that `SpecRunner` receives a `beforeSuiteNode` as a `SuiteNode` interface.  We may need to extend this interface to be able to distinguish between the behavior for `SynchronizedBeforeSuite` vs `BeforeSuite`, etc...

Thanks, again!  And no worries if you don't get to it.  This is a nice-to-have but I don't think the current behavior is especially borked.

**Comment by robdimsdale** [comment]
Thanks for the detailed explanation - it's much appreciated.

It's been a couple of months since I opened this issue and I haven't felt this pain again, so I'm going to close this issue.

If it does become an issue again I may re-open this and work on a PR.
